### PR TITLE
PHPC-536: UTCDateTime construction from DateTime or current time

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPC-536

### CODE REVIEW

**Comment by moderndeveloperllc** [comment]
Sorry for my programming ignorance, but I noticed that you were using `DateTime` in the code and tests. Is this code `DateTimeInterface`-compatible? It may also be advisable to throw in a pre-1970 date in the tests just to make sure that negative numbers are playing nicely. I'm looking forward to this landing soon!

File: src/BSON/UTCDateTime.c [hunk 1]
```diff
@@ -46,17 +51,43 @@ PHONGO_API zend_class_entry *php_phongo_utcdatetime_ce;
...
+			php_date_obj *datetime_obj;
+
+			datetime_obj = Z_PHPDATE_P(datetime);
+			sec = datetime_obj->time->sse;
+			usec = (int64_t) floor(datetime_obj->time->f * 1000000 + 0.5);
+		} else {
+			/* Initialize UTCDateTime from current time */
+			struct timeval cur_time;
+
+			gettimeofday(&cur_time, NULL);
```
Comment (derickr) [inline src/BSON/UTCDateTime.c]: Windows doesn't have ```gettimeofday```.

File: tests/bson/bson-utcdatetime-005.phpt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+$tests = [
+    new DateTime(),
+    new DateTime('@1215282385'),
+    new DateTime('2011-01-01T15:03:01.012345Z'),
+    new DateTime('2050-11-12T13:14:15.999999Z'),
+];
+
+foreach ($tests as $test) {
+    var_dump(new MongoDB\BSON\UTCDateTime($test));
```
Comment (derickr) [inline tests/bson/bson-utcdatetime-005.phpt]: I think you should try to round trip them back with ``toDateTime`` as well.

File: src/BSON/UTCDateTime.c [hunk 3]
```diff
@@ -46,17 +51,43 @@ PHONGO_API zend_class_entry *php_phongo_utcdatetime_ce;
...
+		if (datetime) {
+			/* Initialize UTCDateTime from DateTime object. The following
+			 * assignments use the same logic as date_format() in php_date.c */
+			php_date_obj *datetime_obj;
+
+			datetime_obj = Z_PHPDATE_P(datetime);
+			sec = datetime_obj->time->sse;
+			usec = (int64_t) floor(datetime_obj->time->f * 1000000 + 0.5);
+		} else {
+			/* Initialize UTCDateTime from current time */
```
Comment (derickr) [inline src/BSON/UTCDateTime.c]: Shouldn't this be a different ```zend_parse_parameters_ex```? I think I'd like a case for all three options? But then again, this does work too...

File: tests/bson/bson-utcdatetime-004.phpt [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+BSON BSON\UTCDateTime constructor defaults to current time
+--SKIPIF--
+<?php require __DIR__ . "/../utils/basic-skipif.inc"?>
+--FILE--
+<?php
+require_once __DIR__ . "/../utils/basic.inc";
+
+$tests = [
+    new MongoDB\BSON\UTCDateTime,
+    new MongoDB\BSON\UTCDateTime(null),
```
Comment (derickr) [inline tests/bson/bson-utcdatetime-004.phpt]: I think "null" should just be coerced to ``0``, and not it being empty. Hence, the milliseconds should end up being 0 (e.g., the Epoch)

File: src/BSON/UTCDateTime.c (hunk 1 above)
Comment (jmikola) [inline src/BSON/UTCDateTime.c]: I considered the options in http://stackoverflow.com/q/10905892/162228, but then I found `gettimeofday` and the necessary structs defined in `win32/time.h` (going back to at least PHP 5.4). I also found evidence of that implementation being used in several core extensions.

File: tests/bson/bson-utcdatetime-004.phpt (hunk 4 above)
Comment (jmikola) [inline tests/bson/bson-utcdatetime-004.phpt]: The current behavior is consistent with your [HHVM implementation](https://github.com/mongodb/mongo-hhvm-driver/blob/73bc3daa7691184a6e091585ad42c01cb965fe41/ext_mongodb.php#L829). Note that DateTime's constructor also defaults to the current time if `null` is provided. I say we keep this behavior and document it.

File: tests/bson/bson-utcdatetime-005.phpt (hunk 2 above)
Comment (jmikola) [inline tests/bson/bson-utcdatetime-005.phpt]: Sounds like a good idea. Let's get https://github.com/mongodb/mongo-php-driver/pull/338 merged first in that case, since it fixes `toDateTime()` and the expected output.

File: src/BSON/UTCDateTime.c (hunk 3 above)
Comment (jmikola) [inline src/BSON/UTCDateTime.c]: `zend_parse_parameters()` doesn't have a pattern to pick up `null`, which is equivalent to no argument at all, so this seemed easiest. Otherwise, I would need to change the original `zend_parse_parameters()` calls below (for string/integer parsing) to each allow for `null`.

**Comment by jmikola** [comment]
> Sorry for my programming ignorance, but I noticed that you were using `DateTime` in the code and tests. Is this code `DateTimeInterface`-compatible?

DateTimeInterface only exists in PHP 5.5, but we should be able to use that if it's available and fall back to DateTime for PHP 5.4.

@derick: I noticed that `php_date.h` only exports class entries for DateTime and DateTimeImmutable:

```
PHPAPI zend_class_entry *php_date_get_date_ce(void);
PHPAPI zend_class_entry *php_date_get_immutable_ce(void);
```

Can we get away with defining `extern zend_class_entry * date_ce_interface` and relying on variable in `php_date.c` (it's not declared static). Alternatively, do we need to use `zend_fetch_class()` and resolve the interface name string to a class entry?

> It may also be advisable to throw in a pre-1970 date in the tests just to make sure that negative numbers are playing nicely. I'm looking forward to this landing soon!

Great idea. I'll work on getting these in and try to come up with some tests that exercise the range limits of PHP and BSON dates.

**Comment by jmikola** [comment]
@derickr: Regarding the Travis failure, that appears to be `bson-utcdatetime_error-001.phpt`, which was asserting that `UTCDateTime::__construct()` threw when no argument was passed. Since that test case is no longer valid, I propose we delete it and decrement the other "-002" and "-003" tests.

**Comment by moderndeveloperllc** [comment]
> DateTimeInterface only exists in PHP 5.5, but we should be able to use that if it's available and fall back to DateTime for PHP 5.4.

Sorry. That's just me forgetting that the driver is compatible back to 5.4. Being able to pass in `DateTimeInterface` classes would be great regardless.

File: tests/bson/bson-utcdatetime-004.phpt (hunk 4 above)
Comment (derickr) [inline tests/bson/bson-utcdatetime-004.phpt]: OK

File: src/BSON/UTCDateTime.c (hunk 1 above)
Comment (derickr) [inline src/BSON/UTCDateTime.c]: OK

File: tests/bson/bson-utcdatetime-005.phpt (hunk 2 above)
Comment (derickr) [inline tests/bson/bson-utcdatetime-005.phpt]: OK

File: src/BSON/UTCDateTime.c (hunk 3 above)
Comment (derickr) [inline src/BSON/UTCDateTime.c]: OK

**Comment by derickr** [comment]
> Can we get away with defining extern zend_class_entry * date_ce_interface and relying on variable in php_date.c (it's not declared static).

No. This works on Linux, but won't work on Windows.

> Alternatively, do we need to use zend_fetch_class() and resolve the interface name string to a class entry?

That sounds kinda crappy... I'd rather export the interface in PHP (from PHP 7.1).

**Comment by jmikola** [comment]
>> Alternatively, do we need to use zend_fetch_class() and resolve the interface name string to a class entry?
> That sounds kinda crappy... I'd rather export the interface in PHP (from PHP 7.1).

Of course it makes sense to start doing that in PHP 7.1, but `zend_fetch_class()` seems necessary to get this working today.

File: src/BSON/UTCDateTime.c [hunk 5]
```diff
@@ -47,17 +52,58 @@ PHONGO_API zend_class_entry *php_phongo_utcdatetime_ce;
...
+
+	#if PHP_VERSION_ID >= 70000
+		zend_string *zs_classname = zend_string_init("DateTimeInterface", sizeof("DateTimeInterface")-1, 0);
+		found_ce = zend_fetch_class(zs_classname, ZEND_FETCH_CLASS_AUTO|ZEND_FETCH_CLASS_NO_AUTOLOAD|ZEND_FETCH_CLASS_SILENT);
+		zend_string_free(zs_classname);
+	#else
+		found_ce = zend_fetch_class("DateTimeInterface", sizeof("DateTimeInterface")-1, ZEND_FETCH_CLASS_AUTO|ZEND_FETCH_CLASS_NO_AUTOLOAD|ZEND_FETCH_CLASS_SILENT TSRMLS_CC);
+	#endif
+
+	return found_ce ? found_ce : php_date_get_date_ce();
```
Comment (jmikola) [inline src/BSON/UTCDateTime.c]: @derickr: Would it make sense to memoize this function by saving `found_ce` in a static global? I also thought about initializing the constructors `date_ce` in UTCDateTime's MINIT, but that assumes we are always initialized after PHP's date extension. I'd like to respect the edge case where that may not be true.

File: src/BSON/UTCDateTime.c [hunk 6]
```diff
@@ -47,17 +52,58 @@ PHONGO_API zend_class_entry *php_phongo_utcdatetime_ce;
 
 zend_object_handlers php_phongo_handler_utcdatetime;
 
-/* {{{ proto BSON\UTCDateTime UTCDateTime::__construct(integer $milliseconds)
+static zend_class_entry *php_phongo_utcdatetime_get_date_ce(TSRMLS_D)
+{
+	zend_class_entry *found_ce = NULL;
+
+	#if PHP_VERSION_ID >= 70000
```
Comment (derickr) [inline src/BSON/UTCDateTime.c]: The # needs to be in the first column.

File: src/BSON/UTCDateTime.c (hunk 5 above)
Comment (derickr) [inline src/BSON/UTCDateTime.c]: You can add a dependency on it, and that should make PHP initialise datetime first. Instead of MINIT, I would put it in RINIT.

File: src/BSON/UTCDateTime.c [hunk 7]
```diff
@@ -47,17 +52,60 @@ PHONGO_API zend_class_entry *php_phongo_utcdatetime_ce;
...
 	intern = Z_UTCDATETIME_OBJ_P(getThis());
 
+#if PHP_VERSION_ID >= 50500
+	if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS() TSRMLS_CC, "O", &datetime, php_date_get_immutable_ce()) == SUCCESS) {
+		php_phongo_utcdatetime_init_from_date(intern, Z_PHPDATE_P(datetime));
+
+		zend_restore_error_handling(&error_handling TSRMLS_CC);
+		return;
+	}
+#endif
```
Comment (jmikola) [inline src/BSON/UTCDateTime.c]: I ran into problems calling `zend_fetch_class()` to get the class entry for DateTimeInterface during MINIT. This ultimately seemed an easier solution, which uses nothing more than the date extension's public API.

Meanwhile, I've opened [PHPC-718](https://jira.mongodb.org/browse/PHPC-718) to ensure that we start listing our extension dependencies.

File: src/BSON/UTCDateTime.c [hunk 8]
```diff
@@ -47,17 +52,60 @@ PHONGO_API zend_class_entry *php_phongo_utcdatetime_ce;
...
+		if (datetime) {
+			php_phongo_utcdatetime_init_from_date(intern, Z_PHPDATE_P(datetime));
+		} else {
+			php_phongo_utcdatetime_init_from_current_time(intern);
+		}
+
+		zend_restore_error_handling(&error_handling TSRMLS_CC);
+		return;
+	}
+
```
Comment (derickr) [inline src/BSON/UTCDateTime.c]: I think if you now pass it in as a non datetime class, it will not throw an error/exception.

Comment (jmikola) [inline src/BSON/UTCDateTime.c]: We still end up with an InvalidArgumentException:

> MongoDB\BSON\UTCDateTime::__construct() expects parameter 1 to be long, object given

In the HHVM driver, you end up casting the argument to an integer if it's neither `null` nor a DateTime instance. Would you like to intercept other objects and throw a special message?

Comment (derickr) [inline src/BSON/UTCDateTime.c]: Yes.

Comment (jmikola) [inline src/BSON/UTCDateTime.c]: I went with "Expected instance of DateTimeInterface, $classname given". We technically check for an instance of DateTime and DateTimeImmutable (on 5.5+); however, DateTimeInterface cannot be implemented by userland classes so the message is still accurate.

File: tests/bson/bson-utcdatetime_error-001.phpt [hunk 9]
```diff
@@ -1,22 +1,19 @@
 --TEST--
-BSON BSON\UTCDateTime #001 error
---INI--
-date.timezone=America/Los_Angeles
+BSON BSON\UTCDateTime requires object argument to implement DateTimeInterface
 --SKIPIF--
-<?php if (defined("HHVM_VERSION_ID")) exit("skip HHVM handles parameter parsing differently"); ?>
```
Comment (derickr) [inline tests/bson/bson-utcdatetime_error-001.phpt]: We might have to put this one back in I suppose, if we find that it's still different with HHVM.

**Comment by derickr** [comment]
LGTM

File: tests/bson/bson-utcdatetime_error-001.phpt (hunk 9 above)
Comment (jmikola) [inline tests/bson/bson-utcdatetime_error-001.phpt]: I don't think it will be a problem. Your constructor accepts a mixed type and all validation is left to you, so it should be trivial to customize the exception message here.
